python/ray/autoscaler/v2/instance_manager/node_provider.py

```diff
@@ -1,16 +1,175 @@
 import logging
-from abc import ABCMeta, abstractmethod
-from typing import Dict, List, Set
+from abc import ABC, ABCMeta, abstractmethod
+from dataclasses import dataclass
+from typing import Dict, List, Optional, Set
 
 from ray.autoscaler._private.node_launcher import BaseNodeLauncher
 from ray.autoscaler.node_provider import NodeProvider as NodeProviderV1
 from ray.autoscaler.tags import TAG_RAY_USER_NODE_TYPE
 from ray.autoscaler.v2.instance_manager.config import AutoscalingConfig
+from ray.autoscaler.v2.schema import NodeType
 from ray.core.generated.instance_manager_pb2 import Instance
 
 logger = logging.getLogger(__name__)
 
 
+# Type Alias. This is a **unique identifier** for a cloud instance in the cluster.
+# The provider should guarantee that this id is unique across the cluster,
+# such that:
+#   - When a cloud instance is created and running, no other cloud instance in the
+#     cluster has the same id.
+#   - When a cloud instance is terminated, no other cloud instance in the cluster will
+#     be assigned the same id later.
+CloudInstanceId = str
+
+
+@dataclass
+class CloudInstance:
+    """
+    A class that represents a cloud instance in the cluster, with necessary metadata
+    of the cloud instance.
+    """
+
+    # The cloud instance id.
+    cloud_instance_id: CloudInstanceId
+    # The node type of the cloud instance.
+    node_type: NodeType
+    # Update id from which the cloud instance is launched.
+    update_id: str
+
+
+@dataclass
+class CloudInstanceProviderError:
```

Comment:
why separate errors of CloudInstanceProviderError and LaunchNodeError ?

Reply:
`LaunchNodeError` is one type of the `CloudInstanceProviderError`

Reply:
why not combine them?

Reply:
Oh, so like just `LaunchNodeError` and `TerminateNodeError`? 

Or just `CloudInstanceProviderError` with different causes as maybe like enum?